### ISSUE TITLE
Cannot compile a Wasm module from a called host function using the same `Engine`

### ISSUE THREAD

**Issue by lrubasze** [wasmi-labs/wasmi#631]
The issue is quite similar to #567 
But instead of calling a WASM function we try to
- create another WASM module using the same `Engine`
or
- wrap `Func::wrap` host function for while preinstantiating already created WASM module
```
         lock                                 deadlock
host ------------> WASM #1 ------------> host ------------> create WASM module or wrap `Func::wrap` host function for already created WASM module
```

Would it be possible to fix this?

**Comment by Robbepop** [comment]
This is due to the fact that in `wasmi` the `Engine` is used to allocate function signatures and constants as `EngineResources` instead of the Wasm `Store` for performance reasons. It is simply more efficient for the `Engine` executor to have the resources right away. However, this means that we need to lock the `Engine` while compiling a Wasm module for a bit. And while executing a Wasm function we need to lock it for readability.
We could fix this in several different ways:

1) Move all the stuff into the `Store` which will cause overhead when executing Wasm functions.
2) Make the allocations during Wasm module compilation atomic which would cause overhead for both compilation and execution.
3) Unlock the `EngineResources` before making a call to the host function. This would certainly add overhead for host function calls and maybe even for Wasm function calls.

There might be other solutions.

**Comment by lrubasze** [comment]
Hi @Robbepop!
thanks for quick reply.
The main scenario for us us is the second one in fact (sorry, should have mentioned it in a first place).
Here are more details.
In order to save time we cache compiled WASM modules to be reused later when instantiating another instance. And one of our use cases assumes instantiating from host function invoked from WASM instance.

**Comment by Robbepop** [comment]
As a work around what you could do is the following:

You could resumably call the Wasm functions and instead of compiling a new Wasm module in a host function you instead return an error. Then the resumable function will return and unlock everything while doing so. This allows you to compile the required Wasm module. The only thing left to do is feed back the interrupted Wasm call with the results that the interrupted host function would have returned and resume the Wasm function call.

**Comment by iamyulong** [comment]
Thank you @Robbepop for your swift reply. 

Per your comment, I'd like to add more context for our case.

A little bit background on us, we're from RadixDLT and we've integrated `wasmi v11.0` into [RadixEngine/Scrypto](https://github.com/radixdlt/radixdlt-scrypto) about one year ago, and we've got this problem when attempting to upgrade to `v23.0`.

In RadixEngine, we execute user-provided WASM code dynamically. In order to efficiently execute them, we cache compiled WASM modules. With the new `wasmi` architecture, where `wasmi` module is coupled with engine for performance reason, the idea is broken, because we're not allowed to either 
- A) Add more modules to the engine 
- or B) create new instance of existing module (with separate `Store` for state isolation).

I believe this is applicable for many other Blockchain projects as well.

Sincerely hope you can help prioritise this.

**Comment by Robbepop** [comment]
> As a work around what you could do is the following:
> 
> You could resumably call the Wasm functions and instead of compiling a new Wasm module in a host function you instead return an error. Then the resumable function will return and unlock everything while doing so. This allows you to compile the required Wasm module. The only thing left to do is feed back the interrupted Wasm call with the results that the interrupted host function would have returned and resume the Wasm function call.
> 
>     * `Func::call_resumable`: https://docs.rs/wasmi/0.23.0/wasmi/struct.Func.html#method.call_resumable
> 
>     * `TypedFunc::call_resumable`: https://docs.rs/wasmi/0.23.0/wasmi/struct.TypedFunc.html#method.call_resumable

I'd be delighted to know if my proposed work around solve the issue for you for now until the bug has been fixed in `wasmi`.

**Comment by iamyulong** [comment]
While I like the idea of resumable calls, it's going to be very challenging for our use case.

There are other states that are not resumable in our system. Although the WASM function state can be resumed, all other state updates will have to be aborted.

**Comment by Robbepop** [comment]
There is another work around: Just do what `wasmi` version `0.13.0` (and older) did internally: Spin up a whole new `wasmi::Engine` for each new compiled Module. This would come with plenty of overhead compared to having a single `Engine` but if you come from `wasmi` version `0.11.0` it should at least be on par performance-wise.

**Comment by Robbepop** [comment]
Cloning the `wasmi::Store` won't work because it would only do a shallow-copy of the associated `Engine` and therefore still use the same `wasmi::Engine` under the hood.

**Comment by iamyulong** [comment]
Will it be the same if we do not intent to modify the resources of the Engine, for our specific use case? I viewed resources as the compiled functions, for which I might be wrong. :-)

I saw in https://github.com/paritytech/wasmi/pull/590, you allocate a new stack per `execute_func`. Am I right to assume that you've kind of allowed concurrent function invocation? This is what triggered me to think about reusing instance/module/engine, rather than allocating new `Instance` for already seen code.

**Comment by Robbepop** [comment]
> Will it be the same if we do not intent to modify the resources of the Engine, for our specific use case? I viewed resources as the compiled functions, for which I might be wrong. :-)

Yes, `wasmi::Engine` is always shallow copied. So you simply cannot deeply clone a `wasmi::Engine`.

> I saw in #590, you allocate a new stack per `execute_func`. Am I right to assume that you've kind of allowed concurrent function invocation (within the same engine)? This is what triggered me to think about reusing `(engine, module, instance)` combo, rather than allocating new `Instance` for already seen, parsed and compiled WASM code.

Also yes, `wasmi::Engine` allows to concurrently execute Wasm functions but it cannot concurrently compile Wasm modules. Compiling a Wasm module requires an exclusive handle to the `wasmi::Engine`. This is why you cannot compile a Wasm module from within a called (currently executed) Wasm (or host) function.

**Comment by iamyulong** [comment]
Thanks a lot and that's very cool!

Look forward to concurrent WASM compilation. I guess it will be a bonus for fixing this issue.

And yes, in the above hack approach, we do not intend to clone the engine but the `Store`, so we can leverage the concurrent WASM function execution feature alone (every shallow clone will have a different global, memory and host state). 

Am I right to understand that all "static" states of instance will be in the `Store`?

**Comment by Robbepop** [comment]
> Am I right to understand that all "static" states of instance will be in the Store?

Yes, as per the Wasm specification the `Store` is the owner of all instantiated Wasm related state.

**Comment by Robbepop** [comment]
@lrubasze are you still in need of getting this fixed? I see the original PR was already merged.

**Comment by lrubasze** [comment]
Hi @Robbepop !

yes, we were able to merge PR because at the moment we are using a [hack](https://github.com/radixdlt/wasmi/commit/84085f82bd4a18e6ecee530e1cb16675c7334df7) mentioned [earlier](https://github.com/paritytech/wasmi/issues/631#issuecomment-1400360998).
 
To elaborate more on our approach.
We are caching parsed WASM module "template" Instance (with imports already defined) and Store, which keeps user data. "Template" (Store, Instance) tuple are cached together, and never to be invoked. 
Upon instantiation Instance and Store are cloned, so the state is not shared between instances. So after clone we receive the same `Store`, where we are able to set different data.
As a "side-effect" we got some small performance gain, since instantiate (and define imports) a WASM module just once.
  
Unfortunately since wasmi `v0.24.0` it is no longer possible to easy Clone the store 😄 (`ExternObjectEntity` is not easy cloneable).
So maybe as an alternative you could add support for `Store` cloning?

**Comment by Robbepop** [comment]
> So maybe as an alternative you could add support for `Store` cloning?

Since I want to stick as closely as possible to Wasmtime and Wasmtime's `Store` doesn't implement `Clone` either (probably for the same reasons as `wasmi`) I am not willing to invest into this.

**Comment by Robbepop** [comment]
@lrubasze If you are fine with the massive performance regressions introduced by [the PR](#682) then feel free to fork `wasmi` and merge the PR to fix your original issue. This way you will likely be able to use newer `wasmi` versions until eventually we find a proper fix for the underlying issue without such massive performance regressions.

**Comment by Robbepop** [comment]
Since deferring locks to fix this bug introduced performance regressions way too significant to be of any value I want to show some other alternatives that we could try out.

1) Making `EngineResources` work concurrently under the hood. This is probably going to be a lot of work but it will probably yield better results than deferred locking. However, it will be an especially difficult undertaking since we are bound to very limiting `no_std` environments.
2) Makig `EngineResources` append-only. This way we avoid the problem by using multiple versions of `EngineResources`. Concurrently running executions will use the "old" `EngineResources` if a new Wasm `Module` compilation happend during their execution and changed the underlying `EngineResources`. Instead new executions will use the new `EngineResources` and if no executions using the old `EngineResources` are running, the old `EngineResources` are getting dropped entirely. Not sure if this will work out but it is worth a try and probably a bit simpler than making `EngineResources` work concurrently.

**Comment by lrubasze** [comment]
@Robbepop thanks for update.
Indeed it seems quite complicated to properly fix the problem.
Forgive me being so stubborn on this...
Maybe it is worth considering cloning a `Store` at least as optional feature?
You said it is expensive to clone it. We are not observing this. Could you elaborate more on this? Maybe there is some test/benchmark which could be particularly useful to explore this topic?

**Comment by Robbepop** [comment]
@Irubasze Cloning a `Store` might be cheap in your particular case but in general cloning a `Store` can be a very expensive operation. Looking at the `Store` type internals reveals the amount of work: https://github.com/paritytech/wasmi/blob/master/crates/wasmi/src/store.rs#L73
Note that `Arena<T>` is similar to `Vec<T>`.

In contrast cloning an Instance is comparable to cloning a `u64`, therefore it is always a cheap operation: https://github.com/paritytech/wasmi/blob/master/crates/wasmi/src/instance/mod.rs#L147

I am really not keen on introducing "optional features" since people are going to depend on it and turn this optional feature into one I can never get rid of for the sake of backwards compatibility.

**Comment by lrubasze** [comment]
@Robbepop fair enough. All your arguments sound reasonable indeed. Thanks for your kind support.
Waiting for the proper fix :)

**Comment by Robbepop** [comment]
Today I took another look at this issue and think that a new attempt at allowing to compile Wasm modules in called host functions might be feasible for performance now that [this PR](https://github.com/paritytech/wasmi/pull/694) has been merged.
The former bottleneck was on Wasm function call and return overhead. However, the linked PR mitigates this issue by pulling all the Wasm function call and return logic into the inner execution core and therefore Wasm calls and returns are no longer throttled by this feature.
The only slowdown we might and will see happening is with host function calls themselves. The former PR indicated a performance regression of roughly 90% so that's what we maybe have to deal with.

@lrubasze Are you still in need of this feature?

**Comment by lrubasze** [comment]
@Robbepop thanks for giving this issue another try :)

Yes, we still need need this.

You've mentioned compiling WASM modules in called host functions.
What about the possibility to create another Wasm module using the same Engine? Would it be possible as well?

**Comment by Robbepop** [comment]
Sorry I am not following: what exactly is the difference between compiling and creating a module in your opinion? :)

**Comment by lrubasze** [comment]
Sorry for confusing you.
By "creating a Wasm module" I ment instantiation. As stated in one of previous comments this is the scenario we have:

> In order to save time we cache compiled WASM modules to be reused later when instantiating another instance. And one of our use cases assumes instantiating from host function invoked from WASM instance.

**Comment by Robbepop** [comment]
I have not thought that through but it should be possible to instantiate modules, too, yes.

It seems to me that module instantiation already is no problem, however, setting up a `Linker` with its definitions is. With the change this should be possible, too, as long as there are no infinitely running Wasm executions running concurrently. The above idea is to unblock the `Engine` from modification unless there is an active Wasm execution happening. Host function calls are not considered Wasm execution in that sense. This means though, that if you have a Wasm execution running in an endless loop concurrently you can no longer do any modifcations to the `Engine`. Modifications include module compilation as well as setting up a `Linker`.

**Comment by lrubasze** [comment]
We will give it a try then and let you know.
But please be patient. It might take some time, we've been really busy recently and holiday time is coming..
Is there anything else we need to take care of apart of assuring that change https://github.com/paritytech/wasmi/pull/694 is present?

**Comment by Robbepop** [comment]
You cannot try it out yet, since it has not yet been implemented. Was just a thought that came by.
Of course, you can implement the changes yourself based on the current `master` branch and #694 but I will eventually get to that myself.
Just wanted to let you know about my recent thoughts.

**Comment by lrubasze** [comment]
Oh, from the discussion I had an impression that #694 indirectly implements it (did not take a look in to the PR). We'd better wait for your implementation :) Again, thanks for the feedback.

**Comment by Robbepop** [comment]
> Oh, from the discussion I had an impression that #694 indirectly implements it (did not take a look in to the PR). We'd better wait for your implementation :) Again, thanks for the feedback.

#694 implements it but at the same time is super outdated. Therefore another PR to implement on top of current `master` makes more sense to me.

**Comment by Robbepop** [comment]
No, unfortunately #694 only provided the foundation for this feature to (maybe) be effective.

**Comment by Robbepop** [comment]
I just merged https://github.com/wasmi-labs/wasmi/pull/1116 which allows to instantiate a Wasm module inside a called host function which kinda reminded me of this long-standing issue and gave it some thoughts.

In order to have an acceptable solution we just need to get one thing right: Resolving an already compiled Wasm function from a `CodeMap` must be roughly as fast as with the current locking methodology. (or faster) This is the hot path when executing Wasm.

Today, this hot path just uses a single atomic operation for synchronization per access which is kinda fast.
Other operations are a lot slower on todays `CodeMap`.

1. Compiling an uncompiled Wasm function. (several atomic ops)
2. Initializing an uncompiled Wasm function. (`RwLock`)
3. Setting up a bunch of uninitialized Wasm functions. (`RwLock`)

We can and probably should speed-up 3. significantly by pushing multiple functions at once with a single lock. Special care needs needs to be taken because of this is the only operation that has an effect on the underlying buffer that stores all the Wasm function entities and thus might invalidate pointers.

The 2. case is used during `Module::new` and should be fairly trivial. It won't do a lot besides writing a single function entity data. Today's atomic locking could be used to guard it.

The 1. case already accesses several atomics but it isn't too bad if it is a bit slow since compiling a Wasm function is slow anyways and due to caching and a limit on uncompiled Wasm functions (each Wasm function can only be compiled once and this is only relevant in lazy mode) this operation usually is not on the hot path.

Therefore the main challenge is to make sure that 1) is implemented in a way that the hot-path of accessing the instructions of a pre-compiled Wasm function is as fast as today's mechanism.

**Comment by Robbepop** [comment]
I came up with a smol Gist to display the `CodeMap` API that we ideally want to have disrespecting its current implementation: https://gist.github.com/Robbepop/747a62c4f7ea1288d6e2b95cf4848b54

**Comment by Robbepop** [comment]
#1122 gets us much closer to a future without dead-locks in the Wasmi engine, however, it still does not meet my performance criteria and further research and more implementation work is needed.

**Comment by Robbepop** [comment]
I just merged #1122 which fixes this long-standing issue with this comment: https://github.com/wasmi-labs/wasmi/pull/1122#issuecomment-2221319388
